include/zephyr/bluetooth/uuid.h

```diff
@@ -110,7 +110,7 @@ struct bt_uuid_128 {
  *  @return Pointer to a generic UUID.
  */
 #define BT_UUID_DECLARE_16(value) \
-	((struct bt_uuid *) ((struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))
+	((const struct bt_uuid *) ((const struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))
```

Comment:
Is this and the other 2 breaking changes? 

I.e. if I have an application with 
```c
struct bt_uuid *uuid = BT_UUID_DECLARE_16(VAL);
```

Would this give an compiler error/warning with type mismatch?

Reply:
> this wouldn't do anything, right?

Right

Reply:
> this wouldn't do anything, right?

In this case, the C compiler must assume the const will be cast away, since that is legal to do, so it's not allowed to put the object in ROM.

Reply:
> In this case, the C compiler must assume the const will be cast away, since that is legal to do, so it's not allowed to put the object in ROM.

Isn't that always the case?

The change to `gatt.h` where the `_service` is now cast to a `(void *)` that would similarly throw away the const as well, right? 

So even with the change to `BT_UUID_DECLARE_16`, it seems that we can still just cast the `const` away, so I am not sure I follow the argument here.

Reply:
> The change to gatt.h where the _service is now cast to a (void *) that would similarly throw away the const as well, right?
> 
> 

Actually, service uuid indeed is const, add void * just avoid compiler warning.

Reply:
> Actually, service uuid indeed is const, add void * just avoid compiler warning.

But isn't the warning that you are assigning a const value to a non const target? 

Casting something that is a `const` pointer with `(void *)` effectively makes it non-const. 

Can you share the warning that you are suppressing?

Reply:
> Can you share the warning that you are suppressing?

https://github.com/zephyrproject-rtos/zephyr/pull/66136#discussion_r1415280213

Reply:
As I said in cpmmit at the beginning, this PR is to solve the unreasonable UUID variable and save some memory waste.

As for the warning you mentioned, I really don't understand what's there to be confused about.

Reply:
@LingaoM You haven't shared what the compiler warning is, which is what I am requesting :)

The reason why I'm asking about this, is that the changes to the `BT_UUID_DECLARE` macros may be a breaking API change (given that you are modifying the type of the return value) to something that isn't backwards compatible), and I just want to make sure that these macros should actually be changed.

Reply:
> You haven't shared what the compiler warning is, which is what I am requesting :)

The compiler warning is a type mismatch, a non-const pointer points to a const variable.

Reply:
> macros may be a breaking API change 

If the user still uses non-const pointers, this will indeed generate a warning. I can add this change to the doc to specifically explain it.

Reply:
My ultimate goal is to solve the unreasonableness of the current code. Services uuid and char uuid should not change in the service definition. This is clearly stipulated in the spec, but we define this as a variable. This is very unreasonable. .

Reply:
> @LingaoM You haven't shared what the compiler warning is, which is what I am requesting :)

See https://github.com/zephyrproject-rtos/zephyr/pull/66136/files#r1416594955 and https://github.com/zephyrproject-rtos/zephyr/pull/66136/files#r1416597911 Thanks!

Reply:
> ```c
> #define BT_UUID_DECLARE_16(value) \
> 	((struct bt_uuid *) ((const struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))
> ```

As per my PR, this will give him a clear compile warning if the user tries to point to a const uuid.

But if change this. i am a little worried that this will mislead the user into thinking that this is a variable, thereby mistakenly trying to modify(is it possible?) the UUID and causing an exception.

Reply:
@Thalley Maybe we provided two API:

``` C
#define BT_UUID_DECLARE_16(value) \
	((struct bt_uuid *) ((struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))


#define BT_UUID_DECLARE_CONST_16(value) \
	((const struct bt_uuid *) ((const struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))
```

Reply:
And use `BT_UUID_DECLARE_CONST_16` to replace all `BT_UUID_DECLARE_16 ` usage in Zephyr.

Reply:
To be honest, the `bt_uuid` struct is sort of weird in that regards, since it's not really a variable ever. As you also point out, we don't have (in Zephyr at least) any uses of `BT_UUID_DECLARE` where the resulting variable is modified afterwards. 

The reason why I didn't request changes for this PR is for the above reasons, and I'm only reluctant to use 
```c
#define BT_UUID_DECLARE_16(value) \
	((const struct bt_uuid *) ((const struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))
```
instead of 
```c
#define BT_UUID_DECLARE_16(value) \
	((struct bt_uuid *) ((const struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))
```
Because it does create the compiler warning you've mentioned (not backwards compatible), while not providing any memory optimization compared to the non-breaking suggestion. 

That being said, I will not oppose your suggestion, and you do have a fair point w.r.t. 

> i am a little worried that this will mislead the user into thinking that this is a variable, thereby mistakenly trying to modify(is it possible?) the UUID and causing an exception.

But I think any attempts to modify the resulting pointer would fail miserably.